### PR TITLE
[fix] setup.py requires pyyaml installed

### PR DESCRIPTION
**Pull request by return42**
```
pip install -e .
...
Obtaining file:///usr/local/searx/searx-src
    ERROR: Command errored out with exit status 1:
     command: /usr/local/searx/searx-pyenv/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/usr/local/searx/searx-src/setup.py'"'"'; __file__='"'"'/usr/local/searx/searx-src/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'rn'"'"', '"'"'n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-vzer91m2
         cwd: /usr/local/searx/searx-src/
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/usr/local/searx/searx-src/setup.py", line 10, in <module>
        from searx.version import VERSION_STRING
      File "/usr/local/searx/searx-src/searx/__init__.py", line 19, in <module>
        import searx.settings_loader
      File "/usr/local/searx/searx-src/searx/settings_loader.py", line 8, in <module>
        import yaml
    ModuleNotFoundError: No module named 'yaml'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

### CODE REVIEW

File: utils/searx.sh [hunk 1]
```diff
@@ -333,6 +333,7 @@ git pull
 pip install -U pip
 pip install -U setuptools
 pip install -U wheel
+pip install -U pyyaml
```
Comment (dalf) [inline utils/searx.sh:336]: `pip install -U -e .` should install pyyaml since requirements.txt contains a reference to pyyaml:
https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/requirements.txt#L10

or am I missing something ?

Comment (return42) [reply utils/searx.sh:336]: It is `pip install -r`  ...

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/utils/makefile.python#L125

which installs packages directly from the requirements file.  On the opposite   `pip install -U -e .` will _execute_ the setup function ...

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L22-L23

from the `setup.py` file.  For this pip (or better say [setuptools](https://setuptools.readthedocs.io/en/latest/) ) will execute the `setup.py` file.
  
This is also what happens, when you install from package hosters (PyPi) and not from the source.

Comment (return42) [reply utils/searx.sh:336]: Some additional hints: pip is the command line to get access to a package management and infrastructure.  At the end the package itself is installed by setuptools:

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L4

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L22

If you have all requirements installed, you can also call:

    pyhon setup.py install        # use --help to see options

to install the package.

Usual it is not needed to ship a requirements file in a python package.  We do it here

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L60

just to have one single point of definition. This definition is reused here

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L16-L17

which becomes a part of the setup(...)  arguments here:

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L46

And to build wheels, setuptools need the additional package [wheel](https://wheel.readthedocs.io/en/latest/) ..

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/utils/searx.sh#L505

**Comment by dalf** [comment]
The thing I don't understand is `make docs` works for me and on the CI with Ubuntu 18.04 (but not 20.04).

The log shows that `make pyenvinstall` installs PyYAML as expected, why would it wouldn't be the case ?

```
[alexandre:~/code/searx] master+* ± pip freeze
[alexandre:~/code/searx] master+* 1 ± pyenv which python
/home/alexandre/.pyenv/versions/3.8.6/bin/python
[alexandre:~/code/searx] master+* ± make clean
CLEAN     pyclean
CLEAN     docs-clean
CLEAN     locally installed npm dependencies
CLEAN     intermediate test stuff
CLEAN     clean
[alexandre:~/code/searx] master+* ± make docs
PYENV     usage: source ./local/py3/bin/activate
Collecting pip
  Using cached pip-20.3.3-py2.py3-none-any.whl (1.5 MB)
Collecting wheel
  Using cached wheel-0.36.2-py2.py3-none-any.whl (35 kB)
Collecting setuptools
  Using cached setuptools-51.1.2-py3-none-any.whl (784 kB)
Installing collected packages: pip, wheel, setuptools
  Attempting uninstall: pip
    Found existing installation: pip 20.2.1
    Uninstalling pip-20.2.1:
      Successfully uninstalled pip-20.2.1
  Attempting uninstall: setuptools
    Found existing installation: setuptools 49.2.1
    Uninstalling setuptools-49.2.1:
      Successfully uninstalled setuptools-49.2.1
Successfully installed pip-20.3.3 setuptools-51.1.2 wheel-0.36.2
Collecting babel==2.9.0
  Using cached Babel-2.9.0-py2.py3-none-any.whl (8.8 MB)
Collecting certifi==2020.12.05
  Using cached certifi-2020.12.5-py2.py3-none-any.whl (147 kB)
Collecting flask==1.1.2
  Using cached Flask-1.1.2-py2.py3-none-any.whl (94 kB)
Collecting flask-babel==2.0.0
  Using cached Flask_Babel-2.0.0-py3-none-any.whl (9.3 kB)
Collecting idna==2.10
  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting jinja2==2.11.2
  Using cached Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
Collecting lxml==4.6.2
  Using cached lxml-4.6.2-cp38-cp38-manylinux1_x86_64.whl (5.4 MB)
Collecting pycld3==0.20
  Using cached pycld3-0.20-cp38-cp38-manylinux1_x86_64.whl (13.3 MB)
Collecting pygments==2.1.3
  Using cached Pygments-2.1.3-py2.py3-none-any.whl (755 kB)
Collecting python-dateutil==2.8.1
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting pyyaml==5.3.1
  Using cached PyYAML-5.3.1-cp38-cp38-linux_x86_64.whl
Collecting requests[socks]==2.25.1
  Using cached requests-2.25.1-py2.py3-none-any.whl (61 kB)
Collecting chardet<5,>=3.0.2
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
Collecting click>=5.1
  Using cached click-7.1.2-py2.py3-none-any.whl (82 kB)
Collecting itsdangerous>=0.24
  Using cached itsdangerous-1.1.0-py2.py3-none-any.whl (16 kB)
Collecting MarkupSafe>=0.23
  Using cached MarkupSafe-1.1.1-cp38-cp38-manylinux1_x86_64.whl (32 kB)
Collecting PySocks!=1.5.7,>=1.5.6
  Using cached PySocks-1.7.1-py3-none-any.whl (16 kB)
Collecting pytz>=2015.7
  Using cached pytz-2020.5-py2.py3-none-any.whl (510 kB)
Collecting six>=1.5
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.2-py2.py3-none-any.whl (136 kB)
Collecting Werkzeug>=0.15
  Using cached Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
Installing collected packages: MarkupSafe, Werkzeug, urllib3, pytz, jinja2, itsdangerous, idna, click, chardet, certifi, six, requests, PySocks, flask, babel, pyyaml, python-dateutil, pygments, pycld3, lxml, flask-babel
Successfully installed MarkupSafe-1.1.1 PySocks-1.7.1 Werkzeug-1.0.1 babel-2.9.0 certifi-2020.12.5 chardet-4.0.0 click-7.1.2 flask-1.1.2 flask-babel-2.0.0 idna-2.10 itsdangerous-1.1.0 jinja2-2.11.2 lxml-4.6.2 pycld3-0.20 pygments-2.1.3 python-dateutil-2.8.1 pytz-2020.5 pyyaml-5.3.1 requests-2.25.1 six-1.15.0 urllib3-1.26.2
INFO:searx:load the default settings from /home/alexandre/code/searx/searx/settings.yml
INFO:searx:Initialisation done
build utils/brand.env
PYENV     install .
Obtaining file:///home/alexandre/code/searx
Requirement already satisfied: certifi==2020.12.05 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2020.12.5)
Requirement already satisfied: babel==2.9.0 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.9.0)
Requirement already satisfied: flask-babel==2.0.0 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.0.0)
Requirement already satisfied: flask==1.1.2 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (1.1.2)
Requirement already satisfied: idna==2.10 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.10)
Requirement already satisfied: jinja2==2.11.2 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.11.2)
Requirement already satisfied: lxml==4.6.2 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (4.6.2)
Requirement already satisfied: pygments==2.1.3 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.1.3)
Requirement already satisfied: python-dateutil==2.8.1 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.8.1)
Requirement already satisfied: pyyaml==5.3.1 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (5.3.1)
Requirement already satisfied: requests[socks]==2.25.1 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (2.25.1)
Requirement already satisfied: pycld3==0.20 in ./local/py3/lib/python3.8/site-packages (from searx==0.18.0) (0.20)
Requirement already satisfied: pytz>=2015.7 in ./local/py3/lib/python3.8/site-packages (from babel==2.9.0->searx==0.18.0) (2020.5)
Collecting cov-core==1.15.0
  Using cached cov_core-1.15.0-py3-none-any.whl
Requirement already satisfied: itsdangerous>=0.24 in ./local/py3/lib/python3.8/site-packages (from flask==1.1.2->searx==0.18.0) (1.1.0)
Requirement already satisfied: Werkzeug>=0.15 in ./local/py3/lib/python3.8/site-packages (from flask==1.1.2->searx==0.18.0) (1.0.1)
Requirement already satisfied: click>=5.1 in ./local/py3/lib/python3.8/site-packages (from flask==1.1.2->searx==0.18.0) (7.1.2)
Requirement already satisfied: MarkupSafe>=0.23 in ./local/py3/lib/python3.8/site-packages (from jinja2==2.11.2->searx==0.18.0) (1.1.1)
Collecting linuxdoc==20210110
  Using cached linuxdoc-20210110-py2.py3-none-any.whl (117 kB)
Requirement already satisfied: setuptools in ./local/py3/lib/python3.8/site-packages (from linuxdoc==20210110->searx==0.18.0) (51.1.2)
Collecting mock==4.0.3
  Using cached mock-4.0.3-py3-none-any.whl (28 kB)
Collecting nose2[coverage_plugin]==0.9.2
  Using cached nose2-0.9.2-py2.py3-none-any.whl (137 kB)
Requirement already satisfied: six>=1.7 in ./local/py3/lib/python3.8/site-packages (from nose2[coverage_plugin]==0.9.2->searx==0.18.0) (1.15.0)
Collecting Pallets-Sphinx-Themes==1.2.3
  Using cached Pallets_Sphinx_Themes-1.2.3-py2.py3-none-any.whl (26 kB)
Collecting pycodestyle==2.6.0
  Using cached pycodestyle-2.6.0-py2.py3-none-any.whl (41 kB)
Collecting pylint==2.6.0
  Using cached pylint-2.6.0-py3-none-any.whl (325 kB)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in ./local/py3/lib/python3.8/site-packages (from requests[socks]==2.25.1->searx==0.18.0) (1.26.2)
Requirement already satisfied: chardet<5,>=3.0.2 in ./local/py3/lib/python3.8/site-packages (from requests[socks]==2.25.1->searx==0.18.0) (4.0.0)
Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in ./local/py3/lib/python3.8/site-packages (from requests[socks]==2.25.1->searx==0.18.0) (1.7.1)
```

**Comment by return42** [comment]
> The log shows that make pyenvinstall installs PyYAML as expected, why would it wouldn't be the case ?

The make target do install the pyenv first ..

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/utils/makefile.python#L204-L205

and this will install the requirements directly from the requirements.txt file ..

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/utils/makefile.python#L124-L125

Here you can see, that this is done before setup.py is called.

https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/utils/makefile.python#L191-L192

It may be a little confusing, but we should not mess the setup of the _build environment_ from makefiles with a package installation where all this stuff does not exists.

TLDR: [install_requires vs requirements files](https://packaging.python.org/discussions/install-requires-vs-requirements/) describes the two aspects and why it is not good to pin versions in setup.py .. [donald stufft](https://caremad.io/posts/2013/07/setup-vs-requirement/) describes _abstract_ vs _concrete_ .. but all this goes far beyond / only if you are interested / BTW in practice I do use unpinned versions in booth, which I think is the best compromise to have _one single point of definitions_

**Comment by dalf** [comment]
Thank you for your detail explanation.

I think I understand what is going on (may be, it was obvious for you).

Since PR https://github.com/searx/searx/pull/2408 , `searx.brand` requires to read `searx/settings.yml`  :
https://github.com/searx/searx/blob/5a511f0d620038b8e94c581bcfd3c987082b9414/setup.py#L11

**Review by dalf — APPROVED** [review]
Thank you !
